utils/test_helpers.go

```diff
@@ -1,27 +1,228 @@
 package utils
 
 import (
+	"bytes"
+	"errors"
 	"fmt"
-	fuzz "github.com/AdaLogics/go-fuzz-headers"
+	"os"
 	"reflect"
-	"testing"
+	"runtime"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
 )
 
+type T struct {
+	parent   *T
+	name     string
+	mu       sync.RWMutex
+	skipped  bool
+	failed   bool
+	finished bool
+	cleanups []func()
+	output   bytes.Buffer
+}
+
+// Most of the T functions are copied from the stdlib
+
+var errNilPanicOrGoexit = errors.New("test executed panic(nil) or runtime.Goexit")
+
+type panicHandling int
+
+const (
+	normalPanic panicHandling = iota
+	recoverAndReturnPanic
+)
+
+func (t *T) log(s string) { _, _ = t.output.WriteString(s) }
+func (t *T) Cleanup(f func()) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.cleanups = append(t.cleanups, f)
+}
+func (t *T) Deadline() (deadline time.Time, ok bool) { return time.Time{}, false }
+func (t *T) Log(args ...any) {
+	t.log(fmt.Sprintln(args...))
+}
+func (t *T) Logf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+}
+func (t *T) Error(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.Fail()
+}
+func (t *T) Errorf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.Fail()
+}
+func (t *T) Fatal(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.FailNow()
+}
+func (t *T) Fatalf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.FailNow()
+}
+func (t *T) Skip(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.SkipNow()
+}
+func (t *T) Skipf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.SkipNow()
+}
+func (t *T) Fail() {
+	if t.parent != nil {
+		t.parent.Fail()
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.failed = true
+}
+func (t *T) FailNow() {
```

Comment:
Can we log the stack trace when failing?

Reply:
I could print out the stack trace or save it and print it during cleanup. Do you have a preference?